File: android/src/main/java/dev/eduardoroth/mediaplayer/MediaPlayer.java
package dev.eduardoroth.mediaplayer;

import android.content.ComponentName;
import android.os.Bundle;
import android.util.Log;

import androidx.annotation.OptIn;
import androidx.appcompat.app.AppCompatActivity;
import androidx.fragment.app.Fragment;
import androidx.media3.common.C;
import androidx.media3.common.util.UnstableApi;
import androidx.media3.session.MediaController;
import androidx.media3.session.SessionToken;

import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;
import com.google.common.util.concurrent.ListenableFuture;

import java.io.File;

import dev.eduardoroth.mediaplayer.models.AndroidOptions;
import dev.eduardoroth.mediaplayer.models.ExtraOptions;
import dev.eduardoroth.mediaplayer.models.MediaPlayerNotification;
import dev.eduardoroth.mediaplayer.models.PlacementOptions;
import dev.eduardoroth.mediaplayer.state.MediaPlayerState;
import dev.eduardoroth.mediaplayer.state.MediaPlayerStateProvider;

public class MediaPlayer {
    public static long VIDEO_STEP = 10000;
    private final AppCompatActivity _currentActivity;

    MediaPlayer(AppCompatActivity currentActivity) {
        _currentActivity = currentActivity;
    }

    @OptIn(markerClass = UnstableApi.class)
    public void create(PluginCall call, String playerId, String url, PlacementOptions placement, AndroidOptions android, ExtraOptions extra) {
        Bundle connectionHints = new Bundle();
        connectionHints.putString("playerId", playerId);
        connectionHints.putString("videoUrl", url);
        connectionHints.putSerializable("placement", placement);
        connectionHints.putSerializable("android", android);
        extra.poster = extra.poster != null ? getFinalPath(extra.poster) : null;
        connectionHints.putSerializable("extra", extra);

        SessionToken sessionToken = new SessionToken(
                _currentActivity.getApplicationContext(),
                new ComponentName(_currentActivity.getApplicationContext(), MediaPlayerService.class)
        );
        ListenableFuture<MediaController> futureController = new MediaController
                .Builder(_currentActivity.getApplicationContext(), sessionToken)
                .setConnectionHints(connectionHints)
                .buildAsync();

        futureController.addListener(() -> {
            JSObject ret = new JSObject();
            ret.put("method", "create");
            try {
                _currentActivity.getSupportFragmentManager().beginTransaction().add(R.id.MediaPlayerFragmentContainerView, new MediaPlayerContainer(futureController.get(), playerId), playerId).commit();
                ret.put("result", true);
                ret.put("value", playerId);
                call.resolve(ret);
            } catch (Exception | Error futureError) {
                ret.put("result", false);
                ret.put("message", "An error occurred while creating player with id " + playerId);
                call.resolve(ret);
            }
        }, _currentActivity.getMainExecutor());
    }

    public void play(PluginCall call, String playerId) {
        JSObject ret = new JSObject();
        ret.put("method", "play");
        try {
            MediaPlayerState playerState = MediaPlayerStateProvider.getState(playerId);
            playerState.mediaController.get().play();
            ret.put("result", true);
            ret.put("value", true);
        } catch (Error err) {
            ret.put("result", false);
            ret.put("message", "Player not found");
        }
        call.resolve(ret);
    }

    public void pause(PluginCall call, String playerId) {
        JSObject ret = new JSObject();
        ret.put("method", "pause");
        try {
            MediaPlayerState playerState = MediaPlayerStateProvider.getState(playerId);
            playerState.mediaController.get().pause();
            ret.put("result", true);
            ret.put("value", true);
        } catch (Error err) {
            ret.put("result", false);
            ret.put("message", "Player not found");
        }
        call.resolve(ret);
    }

    public void getDuration(PluginCall call, String playerId) {
        JSObject ret = new JSObject();
        ret.put("method", "getDuration");
        try {
            MediaPlayerState playerState = MediaPlayerStateProvider.getState(playerId);
            long duration = playerState.mediaController.get().getDuration();
            ret.put("result", true);
            ret.put("value", duration == C.TIME_UNSET ? 0 : (duration / 1000));
        } catch (Error err) {
            ret.put("result", false);
            ret.put("message", "Player not found");
        }
        call.resolve(ret);
    }

    public void getCurrentTime(PluginCall call, String playerId) {
        JSObject ret = new JSObject();
        ret.put("method", "getCurrentTime");
        try {
            MediaPlayerState playerState = MediaPlayerStateProvider.getState(playerId);
            long currentTime = playerState.mediaController.get().getCurrentPosition();
            ret.put("result", true);
            ret.put("value", currentTime == C.TIME_UNSET ? 0 : (currentTime / 1000));
        } catch (Error err) {
            ret.put("result", false);
            ret.put("message", "Player not found");
        }
        call.resolve(ret);
    }

    public void setCurrentTime(PluginCall call, String playerId, long time) {
        JSObject ret = new JSObject();
        ret.put("method", "setCurrentTime");
        try {
            MediaController controller = MediaPlayerStateProvider.getState(playerId).mediaController.get();

            long duration = controller.getDuration();
            long currentTime = controller.getCurrentPosition();
            long seekPosition = currentTime == C.TIME_UNSET ? 0 : Math.min(Math.max(0, time * 1000), duration == C.TIME_UNSET ? 0 : duration);
            controller.seekTo(seekPosition);
            ret.put("result", true);
            ret.put("value", seekPosition);
        } catch (Error err) {
            ret.put("result", false);
            ret.put("message", "Player not found");
        }
        call.resolve(ret);
    }

    public void isPlaying(PluginCall call, String playerId) {
        JSObject ret = new JSObject();
        ret.put("method", "isPlaying");
        try {
            MediaPlayerState playerState = MediaPlayerStateProvider.getState(playerId);
            ret.put("result", true);
            ret.put("value", playerState.mediaController.get().isPlaying());
        } catch (Error err) {
            ret.put("result", false);
            ret.put("message", "Player not found");
        }
        call.resolve(ret);
    }

    public void isMuted(PluginCall call, String playerId) {
        JSObject ret = new JSObject();
        ret.put("method", "isMuted");
        try {
            MediaPlayerState playerState = MediaPlayerStateProvider.getState(playerId);
            ret.put("result", true);
            ret.put("value", playerState.mediaController.get().getVolume() == 0);
        } catch (Error err) {
            ret.put("result", false);
            ret.put("message", "Player not found");
        }
        call.resolve(ret);
    }

    public void mute(PluginCall call, String playerId) {
        JSObject ret = new JSObject();
        ret.put("method", "mute");
        try {
            MediaPlayerState playerState = MediaPlayerStateProvider.getState(playerId);
            playerState.mediaController.get().setVolume(0);
            ret.put("result", true);
            ret.put("value", true);
        } catch (Error err) {
            ret.put("result", false);
            ret.put("message", "Player not found");
        }
        call.resolve(ret);
    }

    public void getVolume(PluginCall call, String playerId) {
        JSObject ret = new JSObject();
        ret.put("method", "getVolume");
        try {
            MediaPlayerState playerState = MediaPlayerStateProvider.getState(playerId);
            ret.put("result", true);
            ret.put("value", playerState.mediaController.get().getVolume());
        } catch (Error err) {
            ret.put("result", false);
            ret.put("message", "Player not found");
        }
        call.resolve(ret);
    }

    public void setVolume(PluginCall call, String playerId, Double volume) {
        JSObject ret = new JSObject();
        ret.put("method", "setVolume");
        try {
            MediaPlayerState playerState = MediaPlayerStateProvider.getState(playerId);
            playerState.mediaController.get().setVolume(volume.floatValue());
            ret.put("result", true);
            ret.put("value", volume);
        } catch (Error err) {
            ret.put("result", false);
            ret.put("message", "Player not found");
        }
        call.resolve(ret);
    }

    public void getRate(PluginCall call, String playerId) {
        JSObject ret = new JSObject();
        ret.put("method", "getRate");
        try {
            MediaPlayerState playerState = MediaPlayerStateProvider.getState(playerId);
            ret.put("result", true);
            ret.put("value", playerState.mediaController.get().getPlaybackParameters().speed);
        } catch (Error err) {
            ret.put("result", false);
            ret.put("message", "Player not found");
        }
        call.resolve(ret);
    }

    public void setRate(PluginCall call, String playerId, Double rate) {
        JSObject ret = new JSObject();
        ret.put("method", "setRate");
        try {
            MediaPlayerState playerState = MediaPlayerStateProvider.getState(playerId);
            playerState.mediaController.get().setPlaybackSpeed(rate.floatValue());
            ret.put("result", true);
            ret.put("value", rate);
        } catch (Error err) {
            ret.put("result", false);
            ret.put("message", "Player not found");
        }
        call.resolve(ret);
    }

    public void remove(PluginCall call, String playerId) {
        JSObject ret = new JSObject();
        ret.put("method", "remove");
        try {
            MediaPlayerState state = MediaPlayerStateProvider.getState(playerId);
            state.mediaController.get().stop();
            Fragment playerFragment = _currentActivity.getSupportFragmentManager().findFragmentByTag(playerId);
            if (playerFragment != null) {
                _currentActivity.getSupportFragmentManager().beginTransaction().remove(playerFragment).commit();
            }
            MediaPlayerNotificationCenter.post(MediaPlayerNotification.create(playerId, MediaPlayerNotificationCenter.NOTIFICATION_TYPE.MEDIA_PLAYER_REMOVED).build());
            ret.put("result", true);
            ret.put("value", playerId);
        } catch (Error err) {
            ret.put("result", false);
            ret.put("message", "Player not found");
        }
        call.resolve(ret);
    }

    public void removeAll(PluginCall call) {
        _currentActivity.getSupportFragmentManager().getFragments().forEach(fragment -> {
            String playerId = fragment.getTag();
            _currentActivity.getSupportFragmentManager().beginTransaction().remove(fragment).commit();
            try {
                MediaPlayerState playerState = MediaPlayerStateProvider.getState(playerId);
                playerState.mediaController.get().stop();
            } catch (Error ignored) {
            }
            MediaPlayerNotificationCenter.post(MediaPlayerNotification.create(playerId, MediaPlayerNotificationCenter.NOTIFICATION_TYPE.MEDIA_PLAYER_REMOVED).build());
        });
        JSObject ret = new JSObject();
        ret.put("method", "removeAll");
        ret.put("result", true);
        ret.put("value", "[]");
        call.resolve(ret);
    }

    private String getFinalPath(String url) {
        if (url == null) {
            return null;
        }
        String path = null;
        if (url.startsWith("file:///")) {
            return url;
        } else if (url.startsWith("application") || url.contains("_capacitor_file_")) {
            String filesDir = _currentActivity.getFilesDir() + "/";
            path = filesDir + url.substring(url.lastIndexOf("files/") + "files/".length());
            File file = new File(path);
            if (!file.exists()) {
                Log.e("Media Player", "File not found");
                path = null;
            }
        } else if (url.contains("public/assets")) {
            path = "/android_asset/" + url;
        } else if (url.startsWith("http")) {
            path = url;
        }
        return path;
    }
}

File: ios/Sources/MediaPlayerPlugin/MediaPlayer.swift
import Foundation
import Capacitor
import UIKit
import AVKit

@objc public class MediaPlayer: NSObject {
    var bridge:CAPBridgeProtocol?

    var controllers: [String: MediaPlayerController] = [:]

    @objc func setBridge(bridge: CAPBridgeProtocol) {
        self.bridge = bridge
    }
    
    @objc func create(
        call: CAPPluginCall,
        playerId: String,
        url: URL,
        placement: MediaPlayerPlacementOptions,
        ios: MediaPlayerIosOptions,
        extra: MediaPlayerExtraOptions
    ) {
        DispatchQueue.main.sync {
            guard getMediaPlayerController(playerId: playerId) == nil else {
                call.resolve(["result": false, "method": "create", "message": "Player with playerId \(playerId) is already created"])
                return
            }

            let mediaPlayerController = MediaPlayerController(
                playerId: playerId, url: url, placement: placement, ios: ios, extra: extra
            )
            self.bridge?.viewController?.addChild(mediaPlayerController)
            self.bridge?.viewController?.view.addSubview(mediaPlayerController.view)
            mediaPlayerController.didMove(toParent: self.bridge?.viewController)
            
            self.addMediaPlayerController(playerId: playerId, controller: mediaPlayerController)
            call.resolve(["result": true, "method": "create", "value": playerId]);
        }
    }

    @objc func play(call: CAPPluginCall, playerId: String) {
        guard let controller = getMediaPlayerController(playerId: playerId) else {
            call.resolve(["result": false, "method": "play", "message": "Player with playerId \(playerId) not found"])
            return
        }
        controller.player.play()
        call.resolve(["result": true, "method": "play", "value": playerId])
    }
    @objc func pause(call: CAPPluginCall, playerId: String) {
        guard let controller = getMediaPlayerController(playerId: playerId) else {
            call.resolve(["result": false, "method": "pause", "message": "Player with playerId \(playerId) not found"])
            return
        }
        controller.player.pause()
        call.resolve(["result": true, "method": "pause", "value": playerId])
    }
    @objc func getDuration(call: CAPPluginCall, playerId: String) {
        guard let controller = getMediaPlayerController(playerId: playerId) else {
            call.resolve(["result": false, "method": "getDuration", "message": "Player with playerId \(playerId) not found"])
            return
        }
        call.resolve(["result": true, "method": "getDuration", "value": controller.duration])
    }
    @objc func getCurrentTime(call: CAPPluginCall, playerId: String) {
        guard let controller = getMediaPlayerController(playerId: playerId) else {
            call.resolve(["result": false, "method": "getCurrentTime", "message": "Player with playerId \(playerId) not found"])
            return
        }
        call.resolve(["result": true, "method": "getCurrentTime", "value": CMTimeGetSeconds(controller.player.currentTime())])
    }
    
    @objc func setCurrentTime(call: CAPPluginCall, playerId: String, time: Double) {
        guard let controller = getMediaPlayerController(playerId: playerId) else {
            call.resolve(["result": false, "method": "setCurrentTime", "message": "Player with playerId \(playerId) not found"])
            return
        }
        controller.player.seek(to: CMTimeMake(value: Int64(time*1000), timescale: 1000))
        call.resolve(["result": true, "method": "setCurrentTime", "value": playerId])
    }
    @objc func isPlaying(call: CAPPluginCall, playerId: String) {
        guard let controller = getMediaPlayerController(playerId: playerId) else {
            call.resolve(["result": false, "method": "isPlaying", "message": "Player with playerId \(playerId) not found"])
            return
        }
        call.resolve(["result": true, "method": "isPlaying", "value": controller.player.timeControlStatus == .playing])
    }
    @objc func isMuted(call: CAPPluginCall, playerId: String) {
        guard let controller = getMediaPlayerController(playerId: playerId) else {
            call.resolve(["result": false, "method": "isMuted", "message": "Player with playerId \(playerId) not found"])
            return
        }
        call.resolve(["result": true, "method": "isMuted", "value": controller.player.isMuted])
    }
    @objc func setVisibilityBackgroundForPiP(call: CAPPluginCall, playerId: String, isVisible: Bool){
        guard let controller = getMediaPlayerController(playerId: playerId) else {
            call.resolve(["result": false, "method": "setVisibilityBackgroundForPiP", "message": "Player with playerId \(playerId) not found"])
            return
        }
        DispatchQueue.main.sync {
            if controller.isInPipMode == true {
                controller.view.isHidden = !isVisible
            }
            call.resolve(["result": true, "method": "setVisibilityBackgroundForPiP", "value": controller.view.isHidden])
        }
    }
    @objc func mute(call: CAPPluginCall, playerId: String) {
        guard let controller = getMediaPlayerController(playerId: playerId) else {
            call.resolve(["result": false, "method": "mute", "message": "Player with playerId \(playerId) not found"])
            return
        }
        controller.player.isMuted = true
        call.resolve(["result": true, "method": "mute", "value": true])
    }
    @objc func getVolume(call: CAPPluginCall, playerId: String) {
        guard let controller = getMediaPlayerController(playerId: playerId) else {
            call.resolve(["result": false, "method": "getVolume", "message": "Player with playerId \(playerId) not found"])
            return
        }
        call.resolve(["result": true, "method": "getVolume", "value": controller.player.volume])
    }
    @objc func setVolume(call: CAPPluginCall, playerId: String, volume: Float) {
        guard let controller = getMediaPlayerController(playerId: playerId) else {
            call.resolve(["result": false, "method": "setVolume", "message": "Player with playerId \(playerId) not found"])
            return
        }
        controller.player.volume = volume
        call.resolve(["result": true, "method": "setVolume", "value": volume])
    }
    @objc func getRate(call: CAPPluginCall, playerId: String) {
        guard let controller = getMediaPlayerController(playerId: playerId) else {
            call.resolve(["result": false, "method": "getRate", "message": "Player with playerId \(playerId) not found"])
            return
        }
        call.resolve(["result": true, "method": "getRate", "value": controller.player.rate])
    }
    @objc func setRate(call: CAPPluginCall, playerId: String, rate: Float) {
        guard let controller = getMediaPlayerController(playerId: playerId) else {
            call.resolve(["result": false, "method": "setRate", "message": "Player with playerId \(playerId) not found"])
            return
        }
        controller.player.rate = rate
        call.resolve(["result": true, "method": "setRate", "value": rate])
    }
    @objc func remove(call: CAPPluginCall, playerId: String) {
        guard let controller = getMediaPlayerController(playerId: playerId) else {
            call.resolve(["result": false, "method": "remove", "message": "Player with playerId \(playerId) not found"])
            return
        }
        DispatchQueue.main.sync {
            controller.releasePlayer()
            removeMediaPlayerController(playerId: playerId)
            call.resolve(["result": true, "method": "remove", "value": playerId])
        }
    }
    @objc func removeAll(call: CAPPluginCall) {
        DispatchQueue.main.sync {
            for(_, controller) in controllers{
                controller.releasePlayer()
            }
            removeAllMediaPlayerControllers()
            call.resolve(["result": true, "method": "removeAll", "value": true])
        }
    }

}
